cache/sampleStoreInit.js

```diff
@@ -0,0 +1,197 @@
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * ./cache/sampleStoreInit.js
+ *
+ * Functions for initializing/clearing the redis sample store.
+ */
+'use strict'; // eslint-disable-line strict
+const Sample = require('../db').Sample;
+const Aspect = require('../db').Aspect;
+const featureToggles = require('feature-toggles');
+const redisClient = require('./redisCache').client.sampleStore;
+const samsto = require('./sampleStore');
+const constants = samsto.constants;
+
+/**
+ * Clear all the "sampleStore" keys (for subjects, aspects, samples) from
+ * redis.
+ *
+ * @returns {Promise} upon completion.
+ */
+function eradicate() {
+  const promises = Object.getOwnPropertyNames(constants.indexKey)
+    .map((s) => redisClient.smembersAsync(constants.indexKey[s])
+    .then((keys) => {
+      keys.push(constants.indexKey[s]);
+      return redisClient.delAsync(keys);
+    })
+    .catch((err) => {
+      // NO-OP
+      console.error(err); // eslint-disable-line no-console
+      Promise.resolve(true);
+    }));
+  return Promise.all(promises);
+} // eradicate
+
+/**
+ * Populate the redis sample store with aspects from the db.
+ *
+ * @returns {Promise} which resolves to the list of redis batch responses.
+ */
+function populateAspects() {
+  return Aspect.findAll()
+  .then((aspects) => {
+    const aspectIdx = [];
+    const cmds = [];
+    aspects.forEach((a) => {
+      const key = samsto.toKey(constants.objectType.aspect, a.name);
+      aspectIdx.push(key);
+      cmds.push(['hmset', key, samsto.cleanAspect(a)]);
+    });
+
+    cmds.push(['sadd', constants.indexKey.aspect, aspectIdx]);
+    return redisClient.batch(cmds).execAsync();
+  })
+  .catch(console.error); // eslint-disable-line no-console
+} // populateAspects
+
+/**
+ * Populate the redis sample store with samples from the db.
+ *
+ * @returns {Promise} which resolves to the list of redis batch responses.
+ */
+function populateSamples() {
+  return Sample.findAll()
+  .then((samples) => {
+    const subjectIdx = new Set();
+    const sampleIdx = new Set();
+    const subjectSets = {};
+    const sampleHashes = {};
+
+    samples.forEach((s) => {
+      const nameParts = s.name.split('|');
+
+      // Generate the redis keys for this aspect, sample and subject.
+      const aspKey = samsto.toKey(constants.objectType.aspect,
+        nameParts[1]); // eslint-disable-line no-magic-numbers
+      const samKey = samsto.toKey(constants.objectType.sample, s.name);
+      const subKey = samsto.toKey(constants.objectType.subject,
+        nameParts[0]); // eslint-disable-line no-magic-numbers
+
+      // Track each of these in the master indexes for each object type.
+      sampleIdx.add(samKey);
+      subjectIdx.add(subKey);
+
+      // For creating each individual subject set...
+      if (subjectSets.hasOwnProperty(subKey)) {
+        subjectSets[subKey].push(aspKey);
+      } else {
```

Comment:
just noticed this, while I was running my tests. Considering that we can easily convert an aspect to an aspectKey but not the other way round, the aspect name should be pushed into the subjectSet instead of aspectKey. Correct?